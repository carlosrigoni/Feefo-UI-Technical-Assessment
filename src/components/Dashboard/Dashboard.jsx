import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Tooltip } from "react-tooltip";

export function Dashboard({ uploads, linesAdded, linesSaved }) {
  return (
    <div className="mt-8 bg-white w-full max-w-xl rounded-md m-4 pt-4 shadow-sm">
      <div className="px-4 flex items-center justify-between">
        <div className="flex items-center">
          <FontAwesomeIcon icon="upload" className="mr-2" color="#3eb0eb" />
          <h2 className="font-semibold">Sales</h2>
        </div>
        <FontAwesomeIcon
          icon="info-circle"
          className="mr-1"
          color="#bfbbbb"
          data-tooltip-id="infoTooltip"
        />
        <Tooltip id="infoTooltip" place="top" type="dark" effect="solid">
          Additional sales information
        </Tooltip>
      </div>
      <p className="my-4 text-sm text-gray-500 px-4">
        You had <strong>{uploads} uploads</strong> and <strong>{linesAdded}</strong> lines added.
      </p>
      <div className="grid grid-cols-1 sm:grid-cols-2 border-t">
        <div className="sm:border-r-2 p-4">
          <p className="text-green-600 font-bold text-4xl">{uploads}%</p>
          <p className="text-gray-400 text-sm font-semibold">UPLOAD SUCCESS</p>
        </div>
        <div className="p-4 border-t sm:border-0">
          <p className="text-green-600 font-bold text-4xl">{linesSaved}%</p>
          <p className="text-gray-400 text-sm font-semibold">LINES SAVED</p>
        </div>
      </div>
    </div>
  );
}
