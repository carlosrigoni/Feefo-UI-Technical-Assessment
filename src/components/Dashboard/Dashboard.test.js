import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Dashboard } from "./Dashboard";
import { act } from "react-dom/test-utils";
import { axe } from "jest-axe";

// Mock FontAwesomeIcon since it's not crucial to the test
jest.mock("@fortawesome/react-fontawesome", () => ({
  FontAwesomeIcon: () => <span>Icon</span>,
}));

jest.mock("react-tooltip", () => ({
  Tooltip: ({ children }) => <div>{children}</div>,
}));

describe("Dashboard component", () => {
  const uploads = 10;
  const linesAdded = 20;
  const linesSaved = 30;

  test("renders Sales header", () => {
    render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    expect(screen.getByText("Sales")).toBeInTheDocument();
  });

  test("displays uploads and lines added", () => {
    render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    expect(screen.getByText(/You had/i)).toBeInTheDocument();
    expect(screen.getByText("10 uploads")).toBeInTheDocument();
    expect(screen.getByText("20")).toBeInTheDocument();
  });

  test("displays correct upload success percentage", () => {
    render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    expect(screen.getByText(`${uploads}%`)).toBeInTheDocument();
    expect(screen.getByText("UPLOAD SUCCESS")).toBeInTheDocument();
  });

  test("displays correct lines saved percentage", () => {
    render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    expect(screen.getByText(`${linesSaved}%`)).toBeInTheDocument();
    expect(screen.getByText("LINES SAVED")).toBeInTheDocument();
  });

  test("displays info icon", () => {
    render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    const icons = screen.getAllByText("Icon");
    expect(icons).toHaveLength(2);
  });

  test("displays tooltip with additional sales information", () => {
    render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    expect(
      screen.getByText("Additional sales information")
    ).toBeInTheDocument();
  });

  // Additional tests

  test("matches snapshot", () => {
    const { asFragment } = render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders correctly with zero values", () => {
    render(<Dashboard uploads={0} linesAdded={0} linesSaved={0} />);
    expect(screen.getByText("0")).toBeInTheDocument();
  });

  test("renders correctly with negative values", () => {
    render(<Dashboard uploads={-10} linesAdded={-20} linesSaved={-30} />);
    expect(screen.getByText("-10%")).toBeInTheDocument();
    expect(screen.getByText("-20")).toBeInTheDocument();
    expect(screen.getByText("-30%")).toBeInTheDocument();
  });

  test("renders correctly with large values", () => {
    render(
      <Dashboard uploads={1000000} linesAdded={2000000} linesSaved={3000000} />
    );
    expect(screen.getByText("1000000%")).toBeInTheDocument();
    expect(screen.getByText("2000000")).toBeInTheDocument();
    expect(screen.getByText("3000000%")).toBeInTheDocument();
  });

  test("is accessible", async () => {
    const { container } = render(
      <Dashboard
        uploads={uploads}
        linesAdded={linesAdded}
        linesSaved={linesSaved}
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
