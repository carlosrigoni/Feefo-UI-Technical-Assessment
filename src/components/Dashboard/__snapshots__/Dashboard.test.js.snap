// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-8 bg-white w-full max-w-xl rounded-md m-4 pt-4 shadow-sm"
  >
    <div
      class="px-4 flex items-center justify-between"
    >
      <div
        class="flex items-center"
      >
        <span>
          Icon
        </span>
        <h2
          class="font-semibold"
        >
          Sales
        </h2>
      </div>
      <span>
        Icon
      </span>
      <div>
        Additional sales information
      </div>
    </div>
    <p
      class="my-4 text-sm text-gray-500 px-4"
    >
      You had 
      <strong>
        10 uploads
      </strong>
       and 
      <strong>
        20
      </strong>
       lines added.
    </p>
    <div
      class="grid grid-cols-1 sm:grid-cols-2 border-t"
    >
      <div
        class="sm:border-r-2 p-4"
      >
        <p
          class="text-green-600 font-bold text-4xl"
        >
          10%
        </p>
        <p
          class="text-gray-400 text-sm font-semibold"
        >
          UPLOAD SUCCESS
        </p>
      </div>
      <div
        class="p-4 border-t sm:border-0"
      >
        <p
          class="text-green-600 font-bold text-4xl"
        >
          30%
        </p>
        <p
          class="text-gray-400 text-sm font-semibold"
        >
          LINES SAVED
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
